lib/web_ui/lib/ui_web/src/ui_web/images.dart

```diff
@@ -44,3 +44,10 @@ FutureOr<ui.Image> createImageFromImageBitmap(JSAny imageSource) {
     imageSource as DomImageBitmap,
   );
 }
+
+/// Creates a [ui.Image] from a valid texture source (for example HTMLImageElement | HTMLVideoElement | HTMLCanvasElement).
```

Comment:
The main issue I have here is defining clear ownership semantics. `HTMLImageElement`, `HTMLVideoElement` and `HTMLCanvasElement` are not transferable objects, so this sort of loosely implies that the engine isn't taking ownership of the object. However, the skwasm implementation of this *does* try to take ownership and transfer this object to the raster thread (it doesn't actually do anything different from the above function).

If we want this function to work in a non-owning manner, the skwasm renderer will have to create an image bitmap out of the texture source synchronously and then transfer/use that for the rest of its pipeline.

If we want this function to work in a manner in which the engine takes ownership, it isn't really essentially any different from `createImageFromImageBitmap` except that we're loosening the restriction on the type being passed (i.e. the `domInstanceOfString(imageSource, 'ImageBitmap')` check).

I think there is room for both owning and non-owning APIs. We need to make a decision here on what we want this to do and document it clearly (and treat it consistently with both renderers).

Reply:
I get the ownership transfer problem, but does there need to be two APIs? There aren't two APIs for creating a VideoFrame in the browser for example. If it transfers, once it transfers, it transfers. Since this is a web specific API that is pretty low level, seems like it should be the caller's responsibility to use transferable objects and not touch them again if they are running in SkWasm? In the test for example, there is a different code path for SkWasm that just uses an ImageBitmap instead. Trying to be too prescriptive about the call could limit it's usefulness in practice and take some control out of the caller's hands.

The existing createImageFromBitmapSource has the same problem and doesn't really do anything special about it, why is this API any different? 

Maybe it's purely a documentation problem? It's probably correct to assume that a lot of people won't about the transferability problem if it's not stated clearly in the docs. I don't think there's a good way to check that something is transferable at the JS level that is future proof and throw a better error message in the SKWasm path, but maybe an assert against the known types of transferable objects on the skwasm path plus some info in the docs could help guide people toward using the API properly?

Reply:
The `createImageFromImageBitmap` API specifically says (in its doc comment) that it takes ownership of the object. The caller must assume that the object they passed in is invalid after the call. This is not a "problem", it is just a documented behavior of the API.

It seems like the main impetus behind this change is that you can pass DOM elements (`HTMLImageElement`, `HTMLVideoElement`, `HTMLCanvasElement`, etc) into the API without requiring the user to do a copy step to make it an `ImageBitmap` first. In this case, the API does *not* take ownership of the passed object. The contents of the texture source are expected to be synchronously copied into some internal representation in the renderer, and the caller can expect the source to be valid after the call is over. In the CanvasKit case, it is immediately calling `texImage2D` on the texture source to make a WebGL texture out of it that Skia can use and own internally. The Skwasm side is similar, but the `texImage2D` call actually happens asynchronously on the raster thread. This means in order to have the same semantics, Skwasm must make a copy of the texture source synchronously (via `createImageBitmap` or similar) so that it is (1) transferable and (2) guaranteed to still be around and have the expected contents by the time the web worker starts using it.

The API must have consistent semantics between renderers. If the API accepts an `HTMLVideoElement` when the renderer is CanvasKit, but that same code breaks when using Skwasm, we haven't implemented the API correctly. Users should not have to be aware of which renderer is being used and change their behavior based on that.

Reply:
@eyebrowsoffire so you think it should maybe create an image bitmap in SkWasm if the object it's passed isn't transferable? Then it's just on the user to take a faster path and send a VideoFrame if they don't want the copy to happen. Easy enough to do with some type checking so that bit can be handled transparently.

Reply:
I think it should *unconditionally* create an image bitmap copy in skwasm, regardless of whether it's transferable. The semantics of this API should be that the caller can expect the texture source they passed in to be valid after the call. So even if it is transferable, we should make our own copy so as not to invalidate the texture source the user passed in.

If we need an API where people can efficiently pass objects they don't want ownership of anymore, we should just loosen the restrictions of the `createImageFromImageBitmap` API and possibly rename it to make it more general, since it already has that contract.

Alternatively, maybe this API could just take a boolean argument that's something like `transferOwnership` or something like that, so that the renderer can know whether it's safe to use the object directly or whether it needs to make a copy. Then this API could essentially replace `createImageFromImageBitmap`. What do you think?

Reply:
I like the alternate option of putting the user in control of the transfer. I think that addresses both of our concerns.

Reply:
@eyebrowsoffire added transferOwnership arg with default of false, which will make a copy